File: 2018-02-05-winter/data structure/lecture4/problem4/data/bit/bit.py
#!/usr/bin/python

from random import *

prob = "bit"

cases = [   #
        30,
        1000,
        1000,
        1000,
        100000,
        100000,
        100000,
        100000,
        100000,
        100000,
        100000
        ]
cur = -1
for (n) in cases :
    cout = ""
    cur += 1
    print "make %d..." % cur
##-----
#TODO generate the data

    cout += "%d\n" % n
    for i in range(n) :
        cout += "%d " % randint(1,1000)
    cout += "\n"
    cout += "%d\n" % n
    for i in range(n) :
        opt = choice(["query", "modify"])
        if opt == "query" :
            cout += "query %d\n" % randint(1,n)
        else :
            l = randint(1,n)
            r = randint(l,n)
            d = randint(1,1000)
            cout += "modify %d %d %d\n" % (l, r, d)

##-----
    f = file( prob + str(cur) + ".in", "w" )
    f.write( cout )
    f.close()

File: 2018-02-05-winter/data structure/lecture3/new prob/bit/bit.cpp
#include <cstdio>

const int N = 100000 + 10;

int n, q;
int aa[N];
long long bit[N];

void modify(int r, int delta) {
	for(int i = r; i; i -= i & -i)
		bit[i] += delta;
}
void modify(int l, int r, int delta) {
	modify(r, delta);
	if(l >= 1) modify(l-1, -delta);
}
int query(int pos) {
	long long rt = 0;
	for(int i = pos; i <= n; i += i & -i)
		rt += bit[i];
	return (int)rt;
}
int main() {
	scanf("%d", &n);
	for(int i = 1; i <= n; i++) {
		scanf("%d", aa+i);
		modify(i, i, aa[i]);
	}
	scanf("%d", &q);
	while(q--) {
		char ch[100];
		scanf("%s", ch);
		if(ch[0] == 'q') {
			int pos;
			scanf("%d", &pos);
			printf("%d\n", query(pos));
		} else {
			int l, r, delta;
			scanf("%d%d%d", &l, &r, &delta);
			modify(l, r, delta);
		}
	}
}
